File: frontend/src/lib/api/propr.ts
import { fetchApi } from './core';

// PROPR-1: Propr.xyz prop-firm integration. Every route except /api/propr/status
// 404s while the hidden backend flag (FORVEN_PROPR_ENABLED) is off, and status
// reports only { enabled: false } — the frontend treats any failure as "off".

export interface ProprStatus {
	enabled: boolean;
	allow_live?: boolean;
	api_key_configured?: boolean;
	base_url?: string;
	connected?: boolean;
	connection_error?: string;
	user_id?: string;
	account_id?: string;
	attempt_id?: string;
	attempt_status?: string;
	account_type?: string | null;
	orders_allowed?: boolean;
	closes_allowed?: boolean;
	account_value?: number | null;
	account_error?: string;
}

export interface ProprMirrorCandidate {
	id: string;
	name: string;
	stage: string;
	timeframe?: string | null;
}

export interface ProprMirrorTradeState {
	status: string;
	reason?: string | null;
	strategy?: string;
	asset?: string;
	direction?: string;
	quantity?: number;
	entry_price?: number | null;
	exit_price?: number | null;
	opened_at?: string;
	closed_at?: string;
	recorded_at?: string;
	source_execution_type?: string;
	risk_usd?: number;
	attempts?: number;
	/** RETRY-1: this entry re-armed after a terminal failure while its entry signal stayed live. */
	retry_signal_gated?: boolean;
	retry_rearmed_at?: string;
}

export interface ProprMirrorHalt {
	day?: string;
	day_start_equity?: number;
	equity?: number;
	daily_loss?: number;
	daily_loss_limit_usd?: number;
	daily_halt_at_usd?: number;
	drawdown_used?: number;
	drawdown_allowance_usd?: number;
	drawdown_type?: string;
	starting_balance?: number;
	profit_target_usd?: number | null;
	profit_progress_usd?: number;
	rules_source?: string;
	halted?: boolean;
	reasons?: string[];
	checked_at?: string;
	/** PROPR-ANCHOR-1: how today's day-start equity was determined. */
	anchor_source?: string;
	daily_rule_fully_enforced?: boolean;
}

export interface ProprMirror {
	enabled: boolean;
	strategies: Record<string, string>;
	candidates: ProprMirrorCandidate[];
	state: Record<string, ProprMirrorTradeState>;
	halt?: ProprMirrorHalt;
	/** Orphaned venue positions the mirror found but does not own. */
	unmanaged?: Record<string, unknown>;
	/**
	 * SLICE-1: the challenge account is divided equally across the roster and each
	 * member sizes within its own share. Before this, every member sized off the
	 * FULL balance — six of them could put 2x the account at risk and exhaust the
	 * drawdown allowance in a single session.
	 */
	capital_slice?: {
		roster_size?: number | null;
		challenge_equity_usd?: number | null;
		slice_usd?: number | null;
		reason?: string | null;
		/** MIRROR-RISK-1: the RESOLVED Settings-page risk, whole percent. */
		risk_pct?: number | null;
		risk_usd_per_trade?: number | null;
	} | null;
}

export interface ProprOverview {
	status: ProprStatus;
	positions: Record<string, unknown>[] | null;
	orders: Record<string, unknown>[] | null;
	trades: Record<string, unknown>[] | null;
	attempts: Record<string, unknown>[] | null;
	challenges: Record<string, unknown>[] | null;
	errors?: Record<string, string>;
}

export interface ProprConnectionCheck {
	name: string;
	ok: boolean;
	detail: unknown;
}

export async function getProprEnabled(): Promise<boolean> {
	// The only propr call the sidebar makes; remote=false keeps it instant
	// (no upstream Propr API round-trip just to render the nav).
	try {
		const res = await fetchApi<ProprStatus>('/api/propr/status?remote=false');
		return Boolean(res?.enabled);
	} catch {
		return false;
	}
}

export async function getProprStatus(remote = true): Promise<ProprStatus> {
	return fetchApi(`/api/propr/status?remote=${remote}`);
}

export async function getProprOverview(): Promise<ProprOverview> {
	return fetchApi('/api/propr/overview');
}

export async function setProprApiKey(apiKey: string): Promise<{ ok: boolean; status: ProprStatus }> {
	return fetchApi('/api/propr/api-key', {
		method: 'PUT',
		body: JSON.stringify({ api_key: apiKey })
	});
}

export async function clearProprApiKey(): Promise<{ ok: boolean }> {
	return fetchApi('/api/propr/api-key', { method: 'DELETE' });
}

export async function getProprMirror(): Promise<ProprMirror> {
	return fetchApi('/api/propr/mirror');
}

export async function updateProprMirror(update: {
	enabled?: boolean;
	strategies?: string[];
}): Promise<{ ok: boolean; enabled: boolean; strategies: Record<string, string> }> {
	return fetchApi('/api/propr/mirror', {
		method: 'PUT',
		body: JSON.stringify(update)
	});
}

export async function tickProprMirror(): Promise<{ ok: boolean; result: Record<string, unknown> }> {
	return fetchApi('/api/propr/mirror/tick', { method: 'POST' });
}

export async function closeProprPosition(
	asset: string,
	positionSide: 'long' | 'short',
	quantity: number
): Promise<{ ok: boolean; result: Record<string, unknown> }> {
	return fetchApi('/api/propr/positions/close', {
		method: 'POST',
		body: JSON.stringify({ asset, position_side: positionSide, quantity, confirm: true })
	});
}

export async function cancelProprOrder(
	orderId: string
): Promise<{ ok: boolean; result: Record<string, unknown> }> {
	return fetchApi(`/api/propr/orders/${encodeURIComponent(orderId)}/cancel`, {
		method: 'POST'
	});
}

export async function runProprConnectionTest(): Promise<{
	ok: boolean;
	base_url: string;
	checks: ProprConnectionCheck[];
}> {
	return fetchApi('/api/propr/connection-test', { method: 'POST' });
}

File: forven/routers/propr.py
"""Propr.xyz prop-firm integration API (PROPR-1).

Deliberately hidden: every route except GET /api/propr/status returns 404
while the hidden integration flag (forven.config.propr_enabled) is off, and
status itself reports only {"enabled": false} in that state — a casual caller
learns nothing. The flag is env/config-only (FORVEN_PROPR_ENABLED) and is NOT
in the settings manifest; OPENING a position additionally requires
FORVEN_ALLOW_PROPR_LIVE, while closes, protective legs and cancels need only
the integration flag (PROPR-PERM-2 — see forven/exchange/propr.py).
"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from forven.api_security import require_operator_access
from forven.config import propr_enabled

router = APIRouter(tags=["propr"], dependencies=[Depends(require_operator_access)])


def _require_enabled() -> None:
    if not propr_enabled():
        # 404 (not 403): while the flag is off this surface does not exist.
        raise HTTPException(status_code=404, detail="Not Found")


def _propr():
    from forven.exchange import propr
    return propr


class ApiKeyRequest(BaseModel):
    api_key: str


class MirrorConfigRequest(BaseModel):
    enabled: bool | None = None
    strategies: list[str] | None = None


class ClosePositionRequest(BaseModel):
    asset: str
    position_side: str = Field(pattern="^(long|short)$")
    quantity: float = Field(gt=0)
    confirm: bool = False


@router.get("/api/propr/status")
def propr_status(remote: bool = True):
    return _propr().get_status(include_remote=remote)


@router.put("/api/propr/api-key")
def propr_set_api_key(body: ApiKeyRequest):
    _require_enabled()
    key = str(body.api_key or "").strip()
    if not key:
        raise HTTPException(status_code=400, detail="api_key is required")
    from forven.db import kv_get, kv_set
    from forven.secret_storage import encrypt_secret

    secrets_blob = kv_get("forven:settings:secrets", {}) or {}
    if not isinstance(secrets_blob, dict):
        secrets_blob = {}
    secrets_blob["propr_api_key"] = encrypt_secret(key)
    kv_set("forven:settings:secrets", secrets_blob)
    return {"ok": True, "status": _propr().get_status()}


@router.delete("/api/propr/api-key")
def propr_clear_api_key():
    _require_enabled()
    from forven.db import kv_get, kv_set

    secrets_blob = kv_get("forven:settings:secrets", {}) or {}
    if isinstance(secrets_blob, dict) and "propr_api_key" in secrets_blob:
        secrets_blob.pop("propr_api_key", None)
        kv_set("forven:settings:secrets", secrets_blob)
    return {"ok": True}


@router.get("/api/propr/overview")
def propr_overview():
    """Everything the Propr page renders in one call. Remote reads are
    independent best-efforts — one failing section must not blank the page."""
    _require_enabled()
    propr = _propr()
    overview: dict = {"status": propr.get_status()}
    for key, fn in (
        ("positions", propr.raw_positions),
        ("orders", lambda: propr.list_orders(limit=50)),
        ("trades", lambda: propr.list_trades(limit=50)),
        ("attempts", propr.list_challenge_attempts),
        ("challenges", propr.list_challenges),
    ):
        try:
            overview[key] = fn()
        except Exception as exc:
            overview[key] = None
            overview.setdefault("errors", {})[key] = str(exc)
    return overview


@router.get("/api/propr/mirror")
def propr_get_mirror():
    """Mirror config + per-trade mirror state + strategies the picker offers.

    The mirror is an observer: it copies roster strategies' trades onto the
    Propr account and never touches live/paper execution itself.

    ``unmanaged`` (PROPR-RECONCILE-1) carries the last reconcile's ORPHANED venue
    positions — real leveraged exposure on the challenge account that no mirror
    state entry accounts for (a lost KV write, a hand-placed order). The mirror
    already detects and records them; without surfacing them here the page showed
    a tidy roster while the account carried positions nothing was watching.

    ``halt`` carries ``anchor_source`` / ``daily_rule_fully_enforced`` (PROPR-ANCHOR-1)
    so the challenge-rules panel can say "daily rule only PARTIALLY enforced today"
    on a day whose opening balance we never observed, instead of drawing a gauge
    that implies the venue's daily-loss rule is fully mirrored.
    """
    _require_enabled()
    from forven import propr_mirror

    halt = propr_mirror.get_halt_state()
    # SLICE-1: the challenge account is divided equally across the roster, so the
    # panel can answer "how big will each member trade?" instead of leaving it to
    # be inferred. Derived from the halt state's equity (the same figure the daily
    # rules are measured against) so the two can never disagree on screen.
    capital_slice = None
    try:
        equity = float((halt or {}).get("equity") or 0.0)
        if equity > 0:
            _slice, meta = propr_mirror.mirror_equity_slice(equity)
            # MIRROR-RISK-1: the RESOLVED risk setting, so the page shows the
            # number the next order will actually use (not a hardcoded caption).
            risk_frac = propr_mirror.mirror_risk_fraction()
            meta["risk_pct"] = round(risk_frac * 100.0, 4)
            if meta.get("slice_usd"):
                meta["risk_usd_per_trade"] = round(float(meta["slice_usd"]) * risk_frac, 2)
            capital_slice = meta
    except Exception:  # noqa: BLE001 — a display figure must never break the panel
        capital_slice = None

    return {
        "enabled": propr_mirror.mirror_enabled(),
        "strategies": propr_mirror.mirror_roster(),
        "candidates": propr_mirror.roster_candidates(),
        "state": propr_mirror.get_state(),
        "halt": halt,
        "unmanaged": propr_mirror.get_unmanaged_state(),
        "capital_slice": capital_slice,
    }


@router.put("/api/propr/mirror")
def propr_update_mirror(body: MirrorConfigRequest):
    _require_enabled()
    if body.enabled is None and body.strategies is None:
        raise HTTPException(status_code=400, detail="nothing to update")
    if body.enabled and not _propr().get_api_key():
        raise HTTPException(status_code=400, detail="configure the Propr API key first")
    from forven import propr_mirror

    result = propr_mirror.set_mirror_config(
        enabled=body.enabled, strategy_ids=body.strategies
    )
    return {"ok": True, **result}


@router.post("/api/propr/mirror/tick")
def propr_mirror_tick_now():
    """Force one observer pass (the scheduler runs it every 60s anyway)."""
    _require_enabled()
    from forven.propr_mirror import mirror_tick

    return {"ok": True, "result": mirror_tick()}


@router.post("/api/propr/positions/close")
def propr_close_position(body: ClosePositionRequest):
    """Manual reduce-only close of a Propr position from the page.

    NOT gated by FORVEN_ALLOW_PROPR_LIVE. A close is risk-REDUCING, so it goes
    through the adapter's reduce lane (PROPR-PERM-2), which asks only for the
    integration flag — refusing an exit would strand the position this endpoint
    exists to unwind. Opens are the ones that need the opt-in.
    """
    _require_enabled()
    if not body.confirm:
        raise HTTPException(status_code=400, detail="confirm=true is required")
    side = "sell" if body.position_side == "long" else "buy"
    try:
        result = _propr().close_position(body.asset, body.quantity, side)
    except RuntimeError as exc:
        raise HTTPException(status_code=403, detail=str(exc))
    if isinstance(result, dict) and result.get("error"):
        raise HTTPException(status_code=502, detail=str(result["error"]))
    return {"ok": True, "result": result}


@router.post("/api/propr/orders/{order_id}/cancel")
def propr_cancel_order(order_id: str):
    _require_enabled()
    try:
        result = _propr().cancel_order("", order_id)
    except RuntimeError as exc:
        raise HTTPException(status_code=403, detail=str(exc))
    if isinstance(result, dict) and result.get("error"):
        raise HTTPException(status_code=502, detail=str(result["error"]))
    return {"ok": True, "result": result}


@router.post("/api/propr/connection-test")
def propr_connection_test():
    """Read-only reachability probe: health, auth, attempts, positions."""
    _require_enabled()
    propr = _propr()
    report: dict = {"base_url": propr.get_base_url()}
    checks: list[dict] = []

    def _check(name: str, fn):
        try:
            value = fn()
            checks.append({"name": name, "ok": True, "detail": value})
        except Exception as exc:
            checks.append({"name": name, "ok": False, "detail": str(exc)})

    _check("user", lambda: propr.get_user())
    _check("attempts", lambda: propr.list_challenge_attempts())
    _check("account", lambda: propr.resolve_account(force_refresh=True))
    _check("positions", lambda: propr.raw_positions())
    report["checks"] = checks
    report["ok"] = all(c["ok"] for c in checks)
    return report
